File: 0260-single-number-iii/0260-single-number-iii.py
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        n=len(nums)
        x=nums[0]
        for i in range(1,n):
            x^=nums[i]
        k = (x&(~(x-1)))
        res,ans=0,0
        for i in range(n):
            if(nums[i]&k!=0):
                res^=nums[i]
            else:
                ans^=nums[i]
        return [res,ans]
        

File: 0260-single-number-iii/0260-single-number-iii.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int n = nums.size();
        int x = nums[0];
        for (int i = 1; i < n; i++) {
            x ^= nums[i];
        }

        unsigned int ux = static_cast<unsigned int>(x);
        unsigned int k = ux & -ux;   

        int res = 0, ans = 0;
        for (int i = 0; i < n; i++) {
            if (nums[i] & k) {
                res ^= nums[i];
            } else {
                ans ^= nums[i];
            }
        }

        return {res, ans};
    }
};
